meetings/working-groups/collection-literals/CL-2023-05-10.md

```diff
@@ -0,0 +1,85 @@
+# Collection literals working group meeting for May 10, 2023
+
+## Agenda
+
+* Discuss new options for natural typing of collection literals
+
+## Discussion
+
+Coming out of the previous full LDM on collection literals ([notes](https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-05-03.md#collection-literal-natural-type)), we wanted to explore more options for attaining both flexible usage and ideal performance with a natural type. Two alternative fleshed-out concepts were presented.
+
+### Option 1: anonymous/unspeakable 'collection builder' type
+
+This approach gives the natural type _builder_ semantics with a set of APIs such as `Add` and `AddRange`:
+
+```cs
+var collection = [1, 2, 3];
+collection.Remove(2);
+TakesImmutableArray(collection);
+```
+
+The natural type would be anonymous. The compiler would realize the collection by directly initializing an `ImmutableArray<T>` in the example above, skipping any intermediate bookkeeping that is statically possible to skip, so that the best performance is achieved.
+
+The extra builder concept changes how users would think about naturally-typed collection literals. This could make the mental model more complex. An explicit 'build' called out somehow in syntax might even be preferable over passing the variable implicitly as the targeted collection type.
+
+We briefly considered whether there would be merits to making the type speakable, a new framework collection builder type that could be passed around. It would be yet another collection/builder type to teach people about. If the new builder type was itself capable of doing the building on the stack, it wouldn't be able to be captured or used across awaits. It would also be an advanced type which we wouldn't want normal users to be working with every time they use a naturally-typed collection literals.
+
+#### Conclusion
+
+The working group had some hesitations with this approach.
+
+### Option 2: optimize away `List<T>`
+
+Under this option, the natural type of collection literals would be `List<T>` as previously proposed. A `List<T>` instance would not actually be created if a different backing storage performs better and the difference is not observable by the code.
+
+These performance benefits would extend to all existing code that creates a new `List<T>` instance using syntax. No `List<T>` would be created in code such as the following; instead, this would be lowered to `stackalloc` or similar:
+
+```cs
+List<int> list = new List<int> { 1, 2 };
+foreach (var item in list) // Only usage of 'list'
+{
```

Comment:
```suggestion
{
}
```

Reply:
Would `// ...` on the next line work as well, without implying that the compiler could have disappeared the entire loop as an optimization? I guess an empty loop body isn't that distracting.

Reply:
Main reason I suggested is, the highlighting is getting totally confused after this, and I figured this was the reason :)

Reply:
Oh, great point :D